File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/StringData.java
package xyz.wagyourtail.commons.parsers;

import lombok.val;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;

import java.util.function.Function;

public abstract class StringData<E extends Data.Content> extends Data<String, E> {

    public StringData(String rawContent) {
        super(rawContent);
    }

    public StringData(E content) {
        super(content);
    }

    @Override
    protected String buildRawContent() {
        return getContent().toString();
    }

    @Override
    public String toString() {
        return getRawContent();
    }

    public abstract static class OnlyRaw<E extends Data.Content> extends StringData<E> {
        private final Function<CharReader<?>, E> contentBuilder;

        public OnlyRaw(String rawContent, Function<CharReader<?>, E> contentBuilder) {
            super(rawContent);
            this.contentBuilder = contentBuilder;
        }

        public OnlyRaw(CharReader<?> reader, Function<CharReader<?>, E> contentBuilder) {
            this(contentBuilder.apply(reader).toString(), contentBuilder);
        }

        @Override
        public E getContent() {
            return buildContent();
        }

        @Override
        protected E buildContent() {
            StringCharReader reader = new StringCharReader(getRawContent());
            val content = contentBuilder.apply(reader);
            reader.expectEOS();
            return content;
        }

        protected String buildRawContent() {
            throw new IllegalStateException("raw content should always be present");
        }
    }

    public abstract static class OnlyParsed<E extends Data.Content> extends StringData<E> {

        public OnlyParsed(E content) {
            super(content);
        }

        protected E buildContent() {
            throw new IllegalStateException("content should always be present");
        }
    }

    public static class BuildStringVisitor implements DataVisitor {
        private final StringBuilder sb = new StringBuilder();

        public static String apply(Data<?, ?> data) {
            BuildStringVisitor visitor = new BuildStringVisitor();
            data.accept(visitor);
            return visitor.build();
        }

        @Override
        public boolean visit(Object o) {
            if (!(o instanceof Data<?, ?>)) {
                sb.append(o);
            }
            return true;
        }

        public String build() {
            return sb.toString();
        }
    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/StringData.kt
package xyz.wagyourtail.commonskt.parsers

import xyz.wagyourtail.commonskt.reader.CharReader

abstract class StringData<E: Data.Content> protected constructor(rawContent: String?, content: E?) : Data<String, E>(rawContent, content) {

    constructor(rawContent: String) : this(rawContent, null)

    constructor(content: E) : this(null, content)

    override fun buildRawContent(): String {
        return content.toString()
    }

    override fun toString(): String {
        return rawContent
    }

    abstract class OnlyRaw<E: Content>(rawContent: String) : StringData<E>(rawContent) {

        override val content: E
            get() = buildContent()

        override fun buildContent(): E {
            val reader = CharReader(rawContent)
            val content = checkedBuildContent(reader)
            reader.expectEOS()
            return content
        }

        abstract fun checkedBuildContent(reader: CharReader<*>): E

        override fun buildRawContent(): String {
            throw IllegalStateException("raw content should always be present")
        }

    }

    abstract class OnlyParsed<E: Content>(content: E) : StringData<E>(content) {

        override fun buildContent(): E {
            throw IllegalStateException("content should always be present")
        }

    }

    interface StringDataBuilder<T: StringData<E>, E: Data.Content> {

        operator fun invoke(rawContent: CharReader<*>): T

        fun checked(rawContent: String): T {
            val reader = CharReader(rawContent)
            val data = invoke(reader)
            reader.expectEOS()
            return data
        }

        fun checkedBuildContent(reader: CharReader<*>): E

    }

    class BuildStringVisitor : DataVisitor {
        private val sb = StringBuilder()

        override fun visit(o: Any?): Boolean {
            if (o !is Data<*, *>) {
                sb.append(o)
            }
            return true
        }

        fun build(): String {
            return sb.toString()
        }

        companion object {
            fun apply(data: Data<*, *>): String {
                val visitor = BuildStringVisitor()
                data.accept(visitor)
                return visitor.build()
            }
        }
    }

}